accord-core/src/main/java/accord/utils/async/AsyncChains.java

```diff
@@ -331,6 +331,62 @@ public static <V> AsyncChain<V> failure(Throwable failure)
         return new Immediate<>(failure);
     }
 
+    public static <V, T> AsyncChain<T> map(AsyncChain<V> chain, Function<? super V, ? extends T> mapper, Executor executor)
+    {
+        return chain.flatMap(v -> new Head<T>()
+        {
+            @Override
+            protected void start(BiConsumer<? super T, Throwable> callback)
+            {
+                try
+                {
+                    executor.execute(() -> {
+                        try
+                        {
+                            callback.accept(mapper.apply(v), null);
+                        }
+                        catch (Throwable t)
+                        {
+                            callback.accept(null, t);
+                        }
+                    });
+                }
+                catch (Throwable t)
+                {
+                    callback.accept(null, t);
+                }
+            }
+        });
+    }
+
+    public static <V, T> AsyncChain<T> flatMap(AsyncChain<V> chain, Function<? super V, ? extends AsyncChain<T>> mapper, Executor executor)
+    {
+        return chain.flatMap(v -> new Head<T>()
+        {
+            @Override
+            protected void start(BiConsumer<? super T, Throwable> callback)
+            {
+                try
+                {
+                    executor.execute(() -> {
+                        try
+                        {
+                            mapper.apply(v).addCallback(callback);
+                        }
+                        catch (Throwable t)
+                        {
+                            callback.accept(null, t);
+                        }
+                    });
+                }
+                catch (Throwable t)
+                {
+                    callback.accept(null, t);
```

Comment:
I noticed this pattern in a couple of places so we don't have to resolve it right now, but we should settle on a desired outcome. Either YOLO run the callback or it's better to just not run it and log. Liveness vs correctness tradeoff, but also it errored out in a very unexpected way so how much value is there in running the callback?

Reply:
This is a good point. I think we can probably fix this by using our own Executor interface on which you can submit mutually-exclusive executions in case the executor were providing mutual exclusion. i.e., even if the executor is shutdown, it may permit waiting until it is terminated before taking a mutex and then invoking some continuation.

But let's tag that as a low priority follow-up ticket.

Reply:
added this TODO ` // TODO (low priority, correctness): If the executor is shutdown then the callback may run in an unexpected thread, which may not be thread safe`